envoyauth/response.go

```diff
@@ -343,54 +330,100 @@ func makeHeaderValueOption(k, v string) *ext_core_v3.HeaderValueOption {
 	}
 }
 
-func makeEnvoyHeaderValueOptionsFromHeadersMap(hvo []*ext_core_v3.HeaderValueOption, headers map[string]any) ([]*ext_core_v3.HeaderValueOption, error) {
-	hvo = slices.Grow(hvo, len(headers))
+func (result *EvalResult) getHeadersFromDecision(fieldName string) ([]*ext_core_v3.HeaderValueOption, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, collectEnvoyHeaders)
+}
+
+func (result *EvalResult) getHTTPHeadersFromDecision(fieldName string) (http.Header, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, collectHTTPHeaders)
+}
+
+func getHeadersWithTransformation[T any](
+	decision any,
+	fieldName string,
+	collector func(string, string, *T),
+) (T, error) {
+	var result T
+
+	switch decision := decision.(type) {
+	case bool:
+		return result, nil
+	case map[string]interface{}:
+		headersList, err := extractHeadersFromDecision(decision, fieldName)
+		if err != nil {
+			return result, err
+		}
+
+		for _, headers := range headersList {
+			if err := collectHeaderValues(headers, collector, &result); err != nil {
+				return result, err
+			}
+		}
+		return result, nil
+	default:
+		return result, fmt.Errorf("illegal value for policy evaluation result: %T", decision)
+	}
+}
+
+func extractHeadersFromDecision(decision map[string]interface{}, fieldName string) ([]map[string]interface{}, error) {
+	val, ok := decision[fieldName]
+	if !ok {
+		return nil, nil
+	}
+
+	switch v := val.(type) {
+	case map[string]interface{}:
+		return []map[string]interface{}{v}, nil
+	case []interface{}:
+		headersList := make([]map[string]interface{}, 0, len(v))
+		for _, item := range v {
+			headers, ok := item.(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("type assertion error, expected headers to be of type 'object' but got '%T'", item)
+			}
+			headersList = append(headersList, headers)
+		}
+		return headersList, nil
+	default:
+		return nil, fmt.Errorf("type assertion error, expected headers to be of type 'object' but got '%T'", v)
+	}
+}
+
+func collectHeaderValues[T any](
+	headers map[string]interface{},
+	collector func(string, string, *T),
+	result *T,
+) error {
 	for key, value := range headers {
 		switch val := value.(type) {
 		case string:
-			hvo = append(hvo, makeHeaderValueOption(key, val))
+			collector(key, val, result)
 		case []string:
-			hvo = slices.Grow(hvo, len(val))
```

Comment:
Removing this might yield a small performance hit as the array will need to grow for each sub-element in the Envoy case.